File: rust/bron_kerbosch/src/bron_kerbosch2_gpx.rs
//! Bron-Kerbosch algorithm with pivot of highest degree towards the remaining candidates (IK_GPX)

use crate::bron_kerbosch_pivot::{visit, PivotChoice};
use crate::graph::{connected_vertices, UndirectedGraph, VertexSetLike};
use crate::reporter::Reporter;

pub fn explore<Graph, Rprtr>(graph: &Graph, reporter: &mut Rprtr)
where
    Graph: UndirectedGraph,
    Rprtr: Reporter,
{
    let candidates = connected_vertices(graph);
    if !candidates.is_empty() {
        visit(
            graph,
            reporter,
            PivotChoice::MaxDegree,
            PivotChoice::MaxDegreeLocalX,
            candidates,
            Graph::VertexSet::new(),
            None,
        );
    }
}

File: go/src/bron_kerbosch/bron_kerbosch2_gpx.go
package bron_kerbosch

func bron_kerbosch2_gpx(graph *UndirectedGraph) [][]Vertex {
	// Bron-Kerbosch algorithm with pivot of highest degree within remaining candidates
	candidates := graph.connected_vertices()
	if candidates.IsEmpty() {
		return nil
	}
	excluded := make(VertexSet, len(candidates))
	var reporter SimpleReporter
	visit(graph, &reporter, MaxDegree, MaxDegreeLocal, candidates, excluded, nil)
	return reporter.cliques
}
